File: Rejection-Sampling-PG.py
import random
import math

def input_from_user() -> str:

    while True: # Continues looping until valid word is retrieved from user
        word : str = input("\nEnter a word (all characters are acceptable): ")
        word = word.strip()

        if len(word) == 0:
            print("\nWord must not be left empty.")

        else:
            break

    return word

def initalise_letters (
        letters: dict[str, int],
        word : str
    ) -> None:

    for letter in word:

        if letter not in letters: # Add a new entry to letters if one does not exist
            letters[letter] = 1
        else:
            letters[letter] += 1 # Increment the existing entry in letters

def calculate_num_of_permutations (
        letters: dict[str, int],
        word : str
    ) -> int:

    denominator : int = 1

    for value in letters.values(): # Loop through all values in letters to account for characters that appear more that once and update the denominator accordingly
        denominator *= math.factorial(value)
    
    num_of_permutations : int = math.factorial(len(word)) // denominator # Calculate number of permutations

    return num_of_permutations

def permutation_generator (
        permutations : set[str],
        num_of_permutations : int,
        letters : dict[str, int],
        word : str
    ) -> None:

    while len(permutations) < num_of_permutations: # Begin rejection sampling loop

        letters_copy : dict[str, int] = letters.copy()
        generated_word : list[str] = [""] * len(word) # Initialise generated_word list with placeholder empty strings

        generating_word(generated_word, letters_copy, word)

        permutations.add("".join(generated_word)) # Convert the generated_word list to a string and append the string to permutations

def generating_word (
        generated_word : list[str],
        letters_copy : dict[str, int],
        word : str
    ) -> None:

    used_indexes : list[int] = []

    for letter in word:
    
        # Find a random location to place a letter in the generated_word list
        random_index : int = random.choice([i for i in range(0, len(word)) if i not in used_indexes])
        used_indexes.append(random_index)

        # Select a random letter from the dictionary to place into the generated_word list
        random_letter : str = random.choice([key for key in letters_copy.keys()])
        letters_copy[random_letter] -= 1
    
        if letters_copy[random_letter] == 0: # Remove character from letters_copy if its value is 0
            del letters_copy[random_letter]
    
        generated_word[random_index] = random_letter # Add the random letter to the generated_word list at a random index

def main():

    print("\n\033[4mRejection Sampling Permutation Generator (Python)\033[0m")

    word : str = input_from_user()

    permutations : set[str] = {word} # Permutations are served in a set which automatically handle repeated values
    
    letters : dict[str, int] = dict() # All letters within the user's word are handled through a dictionary (map)

    initalise_letters(letters, word)

    num_of_permutations : int = calculate_num_of_permutations(letters, word)

    print(f"\nThis word has {num_of_permutations} permutation(s).")

    permutation_generator(permutations, num_of_permutations, letters, word)

    print(f"\nAll permutations:\n {permutations}")

if __name__ == "__main__":
    main()

File: Rejection-Sampling-PG.cpp
#include <iostream>
#include <random>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <format>

using namespace std;

random_device rd;
mt19937 gen(rd());

unsigned long long random (
    unsigned long long lower_bound,
    unsigned long long upper_bound
    ) {
    uniform_int_distribution<unsigned long long> dist(lower_bound, upper_bound);
    return dist(gen);
}

void strip(string &str) {

    const string whitespace = " \t\n\r\f\v"; // ASCII representation of whitespace

    const size_t start = str.find_first_not_of(whitespace); // If this method fails to find a character that is not whitespace, then it will return npos

    if (start == string::npos) { // Check if the string is completely empty or full of spaces
        str = "";
    }

    else {
        const size_t end = str.find_last_not_of(whitespace);

        str = str.substr(start, end - start + 1); // // Extract the word from the whitespaces
    }
}

unsigned long long factorial(unsigned long long number) {
    unsigned long long total = 1;

    while (number > 1) {
        total *= number;
        number -= 1;
    }
    return total;
}

void input_from_user(string &word) { // Continues looping until valid word is retrieved from user

    while (true) {

        cout << "\nEnter a word: ";
        getline(cin, word);
        strip(word);

        if (word.empty()) {
            cout << "Word must not be left empty.\n";
        }

        else {
            break;
        }
    }
}

void initalise_letters (
    unordered_map<char, int> &letters,
    string &word,
    vector<char> &unique_keys
    ) {

    for (char letter: word) {

        if (letters.count(letter) == 0) { // Add a new entry to letters if one does not exist
            letters[letter] = 1;
            unique_keys.push_back(letter);
        }
        else {
            letters[letter] += 1; // Increment the existing entry in letters
        }
    }
}

void calculate_num_of_permutations (
    unordered_map<char, int> &letters,
    string &word,
    unsigned long long &num_of_permutations
    ) {

    unsigned long long denominator = 1;

    for (auto value: letters) { // Loop through all values in letters to account for characters that appear more that once and update the denominator accordingly
        denominator *= factorial(value.second);
    }

    num_of_permutations = factorial( word.length()) / denominator; // Calculate number of permutations
}

void generating_word (
    string &word,
    vector<int> &indexes,
    vector<char> &unique_keys_copy,
    unordered_map<char, int> &letters_copy,
    vector<char> &generated_vector
    ) {

    int random_index = 0;
    char random_letter = '\0';
    unsigned long long number_pos = 0;
    unsigned long long letter_pos = 0;

    for (int counter = 0; counter < word.length(); counter++) {

        // Find a random location to place a letter in generated_vector
        number_pos = random(0, (indexes.size() - 1));
        random_index = indexes.at(number_pos);
        indexes.erase(indexes.begin() + number_pos);

        // Select a random letter from the map to place into generated_vector
        letter_pos = random(0, (unique_keys_copy.size() - 1));
        random_letter = unique_keys_copy.at(letter_pos);
        letters_copy.at(random_letter) -= 1;

        if (letters_copy.at(random_letter) == 0) { // Remove character from letters_copy and unique_keys_copy if its value is 0
            letters_copy.erase(random_letter);
            unique_keys_copy.erase(unique_keys_copy.begin() + letter_pos);
        }

        generated_vector.at(random_index) = random_letter; // Add the random letter to the generated_vector at a random index
    }
}

void permutation_generator (
    unordered_set<string> &permutations,
    const unsigned long long &no_of_combinations,
    unordered_map<char, int> &letters,
    vector<char> &unique_keys,
    string &word
    ) {

    while (permutations.size() < no_of_combinations) { // Begin rejection sampling loop

        unordered_map<char, int> letters_copy = letters;
        vector<char> unique_keys_copy = unique_keys;
        vector<char> generated_vector = {};
        string generated_word;
        vector<int> indexes = {};

        for (int i = 0; i < word.length(); i++) { // Initialise generated_vector with placeholder values
            generated_vector.push_back('\0');
            indexes.push_back(i);
        }

        generating_word(word, indexes, unique_keys_copy, letters_copy, generated_vector);

        for (char letter : generated_vector) { // Append all letters in generated_vector to generated_word
            generated_word += letter;
        }

        permutations.insert(generated_word);
    }
}

int main() {

    cout << "\nRejection Sampling Permutation Generator (C++)\n";
    cout << "----------------------------------------------\n";

    string word;

    input_from_user(word);

    unordered_set<string> permutations = {word}; // Permutations are served in a set which automatically handle repeated values

    unordered_map<char, int> letters; // All letters within the user's word are handled through a dictionary (map)

    vector<char> unique_keys = {};

    initalise_letters(letters, word, unique_keys);

    unsigned long long num_of_permutations;

    calculate_num_of_permutations(letters, word, num_of_permutations);

    cout << format("This word has {} permutation(s).\n", num_of_permutations);

    permutation_generator(permutations, num_of_permutations, letters, unique_keys, word);

    cout << "\nAll permutations:\n";

    for (string permutation: permutations) {
        cout << format("{}, ", permutation);
    }
}
